### ISSUE TITLE
U+200C should be treated as word character

### ISSUE THREAD

**Issue by tusooa** [elixir-lang/elixir#12227]
### Elixir and Erlang/OTP versions

Erlang/OTP 25 [erts-13.0.4] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit:ns]

Elixir 1.13.4 (compiled with Erlang/OTP 25)

### Operating system

Gentoo GNU/Linux 5.18.12-gentoo-x86_64

### Current behavior

```
iex(3)> Regex.match?(~r/\W/u, "\u200c")
true
iex(4)> Regex.match?(~r/\w/u, "\u200c")
false
```

### Expected behavior

"\u200c" should match `\w`.

xref: https://git.pleroma.social/pleroma/pleroma/-/issues/2974

**Comment by kipcole9** [comment]
TLDR; I believe `Regex` is working as expected per the documentation.

The Elixir [Regex](https://github.com/elixir-lang/elixir/blob/main/lib/elixir/lib/regex.ex) is built on top of Erlang [re](https://www.erlang.org/doc/man/re.html) module whose documentation say:

> \w
Any character that matches \p{L} or \p{N} matches, plus underscore

Which means the code is executing as expected, but its not necessarily correct.  Looking at the version of `pcre` in OTP 25 we see:
```elixir
iex> Regex.version
{"8.44 2020-02-12", :little}
```
Which is probably quite a bit older than the version you are seeing.

Since `U+200C` is in the general category [Cf](https://www.compart.com/en/unicode/category/Cf) and these are all formatting it could be argued that should be part of the `\w` definition but they aren't.

You could replicate the behaviour you are after with:
```elixir
iex> Regex.match?(~r"[[:Cf:][:N:][:L:]_]", "\u200C")
** (Regex.CompileError) unknown POSIX class name at position 3
    (elixir 1.14.0) lib/regex.ex:231: Regex.compile!/2
    (elixir 1.14.0) expanding macro: Kernel.sigil_r/2
    iex:31: (file)
```
But as you can see, `:re` doesn't support all unicode character categories. I have a library called [unicode_set](https://hex.pm/packages/unicode_set) that does support all the unicode categories so you can do this:
```elixir
iex> Unicode.Regex.match?("[[:Cf:][:N:][:L:]_]", "\u200C")
true
```

**Comment by josevalim** [comment]
Thank you @kipcole9!

@tusooa, you can report this on the Erlang/OTP repo if you desire. But IIRC they keep `\w` restricted on purpose for backwards compatibility (I may be wrong).

**Comment by tusooa** [comment]
```
iex(6)> Regex.match?(~r/[\p{Cf}\p{N}\p{L}_]/u, "\u200c")
true
```
